pysoa/common/transport/asgi/core.py

```diff
@@ -19,73 +22,142 @@
     ASGI_CHANNEL_TYPES,
     ASGI_CHANNEL_TYPE_LOCAL,
     ASGI_CHANNEL_TYPE_REDIS_SENTINEL,
-    ASGI_CHANNEL_TYPES_REDIS,
+    ASGI_CHANNEL_TYPE_REDIS,
 )
 
+logger = logging.getLogger('pysoa.common.transport')
+
 
 def valid_channel_type(instance, attribute, value):
     if not value or value not in ASGI_CHANNEL_TYPES:
         raise ValueError('asgi_channel_type must be one of {}, got {}'.format(ASGI_CHANNEL_TYPES, value))
 
 
+class SentinelMasterConnectionList(object):
+
+    def __init__(self, hosts, redis_kwargs=None, sentinel_refresh_interval=30):
+        if not hosts:
+            raise Exception('hosts cannot be empty')
+        self.hosts = hosts
+        self.sentinel_refresh_interval = sentinel_refresh_interval
+        if redis_kwargs is None:
+            redis_kwargs = {}
+        self.redis_kwargs = redis_kwargs
+        self._master_connection_list = []
+        self._last_sentinel_refresh = 0
+        self._maybe_refresh_masters()
+
+    def _maybe_refresh_masters(self):
+        if (time.time() - self._last_sentinel_refresh) > self.sentinel_refresh_interval:
+            hosts = self._get_master_info()
+            self._master_connection_list = [redis.Redis.from_url(host, **self.redis_kwargs) for host in hosts]
+            self.ring_size = len(hosts)
+
+    def _get_master_info(self):
+        """
+        Get a list of Redis masters from Sentinel. Tries Sentinel hosts until one succeeds; if none succeed,
+        raises a ConnectionError.
+
+        Returns: list of tuples of (host, port)
+        """
+        logger.info('Fetching master list from Sentinel.')
+        master_info_list = []
+        connection_errors = []
+        random.shuffle(self.hosts)
+        for host in self.hosts:
+            try:
+                redis_client = redis.StrictRedis.from_url(host)
+                master_info_list = redis_client.sentinel_masters()
+                break
+            except redis.ConnectionError as e:
+                connection_errors.append('Failed to connect to {}: {}'.format(host, e))
+                continue
+        if not master_info_list:
+            raise ConnectionError('Could not get master info from sentinel\n{}.'.format('\n'.join(connection_errors)))
+        self._last_sentinel_refresh = time.time()
```

Comment:
This is just a style thing, but I think this line would be better located if contained within the `_maybe_refresh_masters` `if` statement.